### ISSUE TITLE
Does not work with Mirage

### ISSUE THREAD

**Issue by reynir** [inhabitedtype/angstrom#118]
The C stubs are not linked into the unikernel. I'm not sure what it takes to fix this, but seeing it's basically a call to `memcpy` and OCaml-internals I figure it should be straightforward to fix if you know how linking in Mirage works.

I attempted to make pure OCaml implementations of the blit functions, but it seems my implementation was unsound since I get `Invalid_argument "index out of bounds"`.

**Comment by reynir** [comment]
In the meantime I fixed the off-by-one error in my code, and now I have a pure OCaml version of angstrom that I can use with MirageOS for now. 

https://github.com/reynir/angstrom/tree/no-c-blit

**Comment by seliopou** [comment]
This was introduced in #103 to fix a performance regression after eliminating cstruct as a dependency. I think there's not much of a difference between the build rules for angstrom and other mirage libraries, so it must be something else. Maybe @yomimono knows what's up?

**Comment by yomimono** [comment]
I'm happy to look at this, but I won't have time to do it properly until early January.  I'm not sure who among the MirageOS maintainers will be more available - perhaps @avsm , @talex5 , or @samoht ? (If you're in this list, it's because I chose the first few people who came to mind for "linking AND mirageOS", filtered by "not attending 34c3").

**Comment by reynir** [comment]
To clarify, the problem seems to not be `memcpy` not being linked, but the angstrom cstubs, e.g.:

    /home/user/.opam/4.05.0+flambda/build/angstrom.0.7.0/_build/default/lib/bigstring.ml:15: undefined reference to `angstrom_bigstring_blit_to_bigstring'

I will try stare at it some further in the future, but it's a bit out of my water.

**Comment by talex5** [comment]
All C code to be used in a Xen unikernel needs to be compiled with the Mini-OS headers rather than the Linux ones. Therefore, C stubs are not linked by default.

Looking around, there's some jbuilder stuff in io-page you could copy:

- https://github.com/mirage/io-page/blob/754b206595d25ff3bc9679b33a186b53f8d31139/lib/jbuild#L15
- https://github.com/mirage/io-page/blob/754b206595d25ff3bc9679b33a186b53f8d31139/lib/jbuild#L19
- https://github.com/mirage/io-page/blob/754b206595d25ff3bc9679b33a186b53f8d31139/META.io-page-xen.template

Looks like it runs `discover.ml` to get the right C flags. Though, alarmingly, it appears to fall back to using the default C flags if pkg-config fails! @djs55 might have more information...

**Comment by seliopou** [comment]
Ok seems straightforward to support. I can't work on this this week but will probably have time next week to get this into the next release. If angstrom suffers from then, the faraday must as well. @reynir if you're motivated to get this working and tested in the mean time, I would be amenable and grateful.

**Comment by martindemello** [comment]
i squashed the changes from @reynir into one commit and fixed a couple of issues:

https://github.com/martindemello/angstrom/commits/no-c-blit

This also trivially fixes #132 - would it be possible to somehow merge this in as a separate opam target? I'm not sure what sort of support ocaml and dune have for this sort of conditional compilation but if you're open to the idea I'll investigate it.

**Comment by seliopou** [comment]
Removing the c stubs would severely affect performance for, as far as I can tell, the majority of the users of the this library. So that's not really an option.

The sort of conditional compilation that `io-page` does is more of an ideal solution to this issue. I'd be happy to consider a fix that's along those lines.

**Comment by martindemello** [comment]
right, i was thiking of conditional compilation to generate a new opam target, say `angstrom-pure-ocaml`

**Comment by seliopou** [comment]
I mispoke when I said conditional compilation. Just having a configuration step that selects the proper compilation/linking flags should be sufficient, which is, I believe, what `io-page` does.

**Comment by avsm** [comment]
There's no need to put in angstrom-specific hacks to work with Mirage.  As Dune is getting improved cross-compilation support in beta17, I'm planning to look into full Mirage cross compilation support using that shortly.  That will provide a comprehensive environment for every application to automatically cross compile C stubs a link into a unikernel, rather than having to apply workarounds in individual projects.

**Comment by cfcs** [comment]
@avsm is there a timeline on that?

**Comment by seliopou** [comment]
cc @diml

**Comment by diml** [comment]
yes?

**Comment by seliopou** [comment]
Is there a timeline for this? If not or it's sufficiently far out then project maintainers should take independent action I think.

**Comment by diml** [comment]
For cross-compilation in Dune? It's already supported and released in opam

**Comment by seliopou** [comment]
Ah, so it's just about getting mirage-specific support in? If that's not built into Dune, wouldn't that still then require per-project support? I'm confused. @avsm what do you have in mind?

**Comment by avsm** [comment]
I have no timeline at this stage beyond beginning to port the mirage C libraries into a scheme similar to opam-cross.  I'm also pretty busy due to various deadlines for the next few weeks, so if you need to take independent action to work around this, then please do so.  There are a number of other hacks all over the place to systematically fix, so I expect this to take a few months to sort out unless someone can work on it fulltime.

**Comment by cfcs** [comment]
OK - it sounds like it's a bit in the distant future, so for the foreseeable future the way to run Angstrom on Mirage is "pin @reynir's branch and take a 40% performance hit" - while it's super nice that there's a way to do it, it's probably not the most sustainable distribution method. :)

@seliopou, @abeaumont has been using David (@pqwy) Kaloper's `ocb-stubblr` magic (magic, it magically just takes care of it), which I believe he wrote to solve this issue when having to compile C stubs for `nocrypto` (for Mirage). There's a quite clean application of it here (using `ocamlbuild`): https://github.com/abeaumont/ocaml-chacha (see `myocamlbuild.ml` and `pkg/topkg.ml`).

So I don't know, the only "easy" ways forward for Mirage support are one of these:
- making an OPAM package like `angstrom-but-slower-but-without-c-dependencies` that the Mirage crowd could use until Jane Street implements something like ocb-stubblr
  - or a `angstrom-just-as-fast-but-with-a-dependency-on-cstruct` :-)
- switching the build system (can always revert that change when Dune catches up)

Thoughts?

**Comment by seliopou** [comment]
The build hack that @talex5 originally mentioned seems like a fine way to go about it until jbuilder catches up.

**Comment by dinosaure** [comment]
I'm not really happy about this issue and, as I understand correctly all of history, it's specifically because `angstrom` wants to delete the `cstruct` dependency than this issue appeared (with some noises like switching to jbuilder/dune).

And from #1, I continue to not understand why it's really important to don't have `cstruct` dependency. The reason seems to be avoid checks from `cstruct` and _"be more fast"_, but IMHO, it's better to keep security guards than be faster than _something else_ ... (I said _"something else"_ because, at this stage, it could be difficult to find a parser combinator library more efficient than `angstrom` from what I know).

Finally, I can understand story about `memcpy` (and it seems the initial problem) but `cstruct` provides this implementation already, and it's not really clear than an OCaml's `memcpy` will be a large problem about performance - for example, in `decompress`, `memcpy` is required (but about semantic than performance) but it's clearly not the bottleneck about performance (which is the ADLER-32 checksum).

I don't want to apply `decompress`'s mind on angstrom (which is irrelevant) but I just would like to show some obscures corner-parts and the status of this issue - and if it's really relevant to not have `cstruct` as a dependency.

Then, I really would like to fix this problem because `angstrom` come step by step as a deep dependency (for `ocaml-git`) and I can not continue to fix API and waiting a probable fix about MirageOS compilation.

**Comment by diml** [comment]
I haven't followed all the details, but IIUC, this is about having C stubs and be compatible with mirage when using jbuilder? If that's the case, then it works fine with jbuilder, we have something like this in bin_prot, see [the xen folder](https://github.com/janestreet/bin_prot/tree/master/xen) and [the META template](https://github.com/janestreet/bin_prot/blob/master/META.bin_prot.template)

**Comment by seliopou** [comment]
Yeah I've just been waiting for somebody that knows how the linking hacks
work (or cares to learn) to put them in a PR and ensure they're tested.

**Comment by diml** [comment]
Looking at bin_prot, I think adding something like this to `lib/jbuild` should be enough:

```
(library
 ((name        angstrom_xen)
  (public_name angstrom.xen)
  (optional)
  (modules     ())
  (libraries   (bigarray mirage-xen-ocaml))
  (c_names     (bigstring_stubs_xen))
  (c_flags     (:include clfags-xen.sexp))))

(rule (copy# bigstring_stubs.c bigstrings_stubs_xen.c))

(rule
 (with-stdout-to cflags-xen.sexp
  (progn
   (echo "(")
   (run pkg-config pkg-config --static mirage-xen --cflags)
   (echo ")"))))
```

and to `META.angstrom.template`:

```
xen_linkopts = "-langstrom_xen_stubs"
```

**Comment by seliopou** [comment]
While benchmarking httpaf I noticed that there were still some bigarray operations at the top of the perf, and their source was faraday. Rather than duplicating the bigstring code in angstrom, and along with that duplicating this issue, I've created a new library called bigstringaf that both angstrom and faraday will depend on for bigstring operations.

So I'm going to close this issue but I've opened inhabitedtype/bigstringaf#1. I'd still appreciate somebody with access to a xen box to adapt @diml's fix, test it out and send in a PR to bigstringaf. I'd be happy to merge it.
